wp/plugin/lib/equality.ml

```diff
@@ -0,0 +1,387 @@
+      | Some label1, Some label2 -> compare_lbl label1 label2 map
+      | None, None -> Some map
+      | _, _ -> None
+    end
+  | Ret label1, Ret label2 -> compare_lbl label1 label2 map
+  | Int (code1, _tid1), Int (code2, _tid2) -> if code1 = code2 then Some map else None
+  | _, _ -> None
+
+(* There should be no phi terms. Fail if there are. *)
+let compare_phis (phis1 : (phi term) list) (phis2 : (phi term) list)
+    (map : varToVarMap) : varToVarMap option =
+  if List.length phis1 > 0 || List.length phis2 > 0 then None
+  else Some map
+
+(* Check that all defs in two lists match in expression and structure but not TID. *)
+let compare_defs (defs1 : (def term) list)
+    (defs2 : (def term) list) (map : varToVarMap) : varToVarMap option =
+  match List.zip defs1 defs2 with
+  | Core_kernel.List.Or_unequal_lengths.Unequal_lengths -> None
+  | Core_kernel.List.Or_unequal_lengths.Ok z ->
+    List.fold z ~init:(Some map)
+      ~f:(fun (map_wrap) (d1, d2) ->
+          match map_wrap with
+          | None -> None
+          | Some map -> compare_def d1 d2 map)
+
+(* Check that all jmps in a list match in expression and structure but not TID. *)
+let compare_jmps (jmps1 : (jmp term) list) (jmps2 : (jmp term) list)
+    (map : varToVarMap) : varToVarMap option =
+  match List.zip jmps1 jmps2 with
+  | Core_kernel.List.Or_unequal_lengths.Unequal_lengths -> None
+  | Core_kernel.List.Or_unequal_lengths.Ok z ->
+    List.fold z ~init:(Some map)
+      ~f:(fun (map_wrap) (j1, j2) ->
+          match map_wrap with
+          | Some map -> compare_jmp j1 j2 map
+          | None -> None)
+
+(* Compares everything about a block EXCEPT tids, which will be done later. *)
+let compare_block (blk1 : blk term) (blk2 : blk term) map =
+  compare_defs (Term.enum def_t blk1 |> Sequence.to_list)
+    (Term.enum def_t blk2 |> Sequence.to_list) map >>=
+  fun map -> compare_jmps (Term.enum jmp_t blk1 |> Sequence.to_list)
+    (Term.enum jmp_t blk2 |> Sequence.to_list) map >>=
+  fun map -> compare_phis (Term.enum phi_t blk1 |> Sequence.to_list)
+    (Term.enum phi_t blk2 |> Sequence.to_list) map
+
+(* Iterate through all reachable nodes in each cfg
+ *  and generate a map from tid to blk.*)
+let get_node_maps
+    (cfg1 : Bap.Std.Graphs.Ir.t)
+    (cfg2 : Bap.Std.Graphs.Ir.t) : ((blk term) TidMap.t) * ((blk term) TidMap.t) =
+  let acc = fun ele tid_map ->
+    let tid = Term.tid (get_label ele) in
+    TidMap.change tid_map tid ~f:(fun _ -> Some (get_label ele)) in
+  let tid1_map =
+    BFS.fold acc TidMap.empty cfg1 in
+  let tid2_map =
+    BFS.fold acc TidMap.empty cfg2 in
+  tid1_map, tid2_map
+
+(* Compares a sub to another sub; returns a map from tid into sub1
+   to the set of sub2 tids that it is syntactically equal to. Also returns
+   a map from (tid_blk_sub1, tid_blk_sub2) to the mapping between variables
+   used in those subs. *)
+let compare_blocks (sub1: Sub.t) (sub2 : Sub.t) :
+  (TidSet.t TidMap.t) * (varToVarMap TidTupleMap.t) =
```

Comment:
I have a suspicion that it would be better to use `((Var.t Var.Map.t)  Tid.Map.t) Tid.Map.t` rather than `varToVarMap TidTupleMap.t`. Essentially, currying the tid keys in that dictionary

Reply:
Fun stuff.